pkg/autoupdate/tuf/library_manager.go

```diff
@@ -0,0 +1,389 @@
+	// Check to see if the current running version is the version we were requested to add;
+	// return early if it is, but don't error out if we can't determine the current version.
+	currentVersion, err := ulm.currentRunningVersion(binary)
+	if err != nil {
+		level.Debug(ulm.logger).Log("msg", "could not get current running version", "binary", binary, "err", err)
+	} else if currentVersion.Original() == ulm.versionFromTarget(binary, targetFilename) {
+		// We don't need to download the current running version because it already exists,
+		// either in this updates library or in the original install location.
+		return nil
+	}
+
+	if ulm.alreadyAdded(binary, targetFilename) {
+		return nil
+	}
+
+	// Remove downloaded archives after update, regardless of success
+	defer ulm.tidyStagedUpdates(binary)
+
+	stagedUpdatePath, err := ulm.stageUpdate(binary, targetFilename)
+	if err != nil {
+		return fmt.Errorf("could not stage update: %w", err)
+	}
+
+	if err := ulm.verifyStagedUpdate(binary, stagedUpdatePath); err != nil {
+		return fmt.Errorf("could not verify staged update: %w", err)
+	}
+
+	if err := ulm.moveVerifiedUpdate(binary, targetFilename, stagedUpdatePath); err != nil {
+		return fmt.Errorf("could not move verified update: %w", err)
+	}
+
+	return nil
+}
+
+// alreadyAdded checks if the given target already exists in the update library.
+func (ulm *updateLibraryManager) alreadyAdded(binary autoupdatableBinary, targetFilename string) bool {
+	updateDirectory := filepath.Join(ulm.updatesDirectory(binary), ulm.versionFromTarget(binary, targetFilename))
+
+	return autoupdate.CheckExecutable(context.TODO(), executableLocation(updateDirectory, binary), "--version") == nil
+}
+
+// versionFromTarget extracts the semantic version for an update from its filename.
+func (ulm *updateLibraryManager) versionFromTarget(binary autoupdatableBinary, targetFilename string) string {
+	// The target is in the form `launcher-0.13.6.tar.gz` -- trim the prefix and the file extension to return the version
+	prefixToTrim := fmt.Sprintf("%s-", binary)
+
+	return strings.TrimSuffix(strings.TrimPrefix(targetFilename, prefixToTrim), ".tar.gz")
+}
+
+// stageUpdate downloads the update indicated by `targetFilename` and stages it for
+// further verification.
+func (ulm *updateLibraryManager) stageUpdate(binary autoupdatableBinary, targetFilename string) (string, error) {
+	stagedUpdatePath := filepath.Join(ulm.stagedUpdatesDirectory(binary), targetFilename)
+	out, err := os.Create(stagedUpdatePath)
+	if err != nil {
+		return "", fmt.Errorf("could not create file at %s: %w", stagedUpdatePath, err)
+	}
+	defer out.Close()
+
+	resp, err := ulm.mirrorClient.Get(ulm.mirrorUrl + fmt.Sprintf("/kolide/%s/%s/%s", binary, runtime.GOOS, targetFilename))
+	if err != nil {
+		return stagedUpdatePath, fmt.Errorf("could not make request to download target %s: %w", targetFilename, err)
+	}
+	defer resp.Body.Close()
+
+	_, err = io.Copy(out, resp.Body)
```

Comment:
I think I was modeling it off the old code, yeah! Looking at how go-tuf does it -- they write to a file (os.File fulfilling the [Destination interface](https://github.com/theupdateframework/go-tuf/blob/master/client/client.go#L868-L871)) that is also deleted if verification fails: https://github.com/theupdateframework/go-tuf/blob/master/client/client.go#L873. It is a little more efficient because they're writing while generating the metadata at the same time -- I'll take a look at switching to something similar!

Reply:
I'm coming at it thinking that writing and deleting feels like extra steps, potential for errors, and leaving known-bad data around. I'd probably dump the body into an io.Tee, checksum and buffer. If the checksum passes, then we can write it to disk and move on.

Though that Desintaition interface is cute too. But I don't think we need the flexibility.

Reply:
Just seeing this after going with a slightly different implementation that mostly leans on go-tuf -- 

```
	// Request download from mirror
	resp, err := ulm.mirrorClient.Get(ulm.mirrorUrl + fmt.Sprintf("/kolide/%s/%s/%s", binary, runtime.GOOS, targetFilename))
	if err != nil {
		return stagedUpdatePath, fmt.Errorf("could not make request to download target %s: %w", targetFilename, err)
	}
	defer resp.Body.Close()

	// Wrap the download in a LimitReader so we download at most localMeta.Length bytes
	stream := io.LimitReader(resp.Body, localTargetMetadata.Length)

	// Read the target file, simultaneously writing it to the staging directory and generating its metadata
	actualTargetMeta, err := util.GenerateTargetFileMeta(io.TeeReader(stream, out), localTargetMetadata.HashAlgorithms()...)
	if err != nil {
		return stagedUpdatePath, fmt.Errorf("could not write downloaded target %s to file %s and compute its metadata: %w", targetFilename, stagedUpdatePath, err)
	}

	// Verify the actual download against the confirmed local metadata
	if err := util.TargetFileMetaEqual(actualTargetMeta, localTargetMetadata); err != nil {
		return stagedUpdatePath, fmt.Errorf("verification failed for target %s staged at %s: %w", targetFilename, stagedUpdatePath, err)
	}
```

I'll go ahead and add a buffer instead of directly teeing it to `out`, but otherwise I think this addresses what you're suggesting!

Reply:
Updated ✅